### DISCUSSION TITLE
Registering observers with the ObservedBy attribute does not work as expected

### DISCUSSION THREAD

**Discussion by sdkakcy** [laravel/framework#53607] [Ideas]
### Laravel Version

10.48.17

### PHP Version

8.2

### Description

Greetings,
In my project I use the following package to make my models archivable: [Laravel Archivable](https://github.com/joelbutcher/laravel-archivable)
This package fires 4 new events to be used in the corresponding observer class. There is no problem using EventServiceProvider's `boot()` method to call model's `observe` method or the `$observers` property to register the observer, but it does not work as expected when using the `ObservedBy` attribute. This is because when the `HasEvents` trait is booted, the observer is registered at that moment. Therefore, even if another bootable trait in the model calls the `addObservableEvents` method in the initialize method, it will have no effect. Because this operation is done after HasEvents has been booted, the observer is already registered.

### Steps To Reproduce

- Create a model with an Observer
- Create a bootable trait with boot and initialize methods and use it on model
- Call `addObservableEvents` with custom model event names in initialize method
- Add ObservedBy attribute to your model for register observer
- Take an action that fires the special event you have created

---

<img width="1048" alt="Screenshot 2024-11-20 at 19 37 38" src="https://github.com/user-attachments/assets/219eb569-db8d-4b63-a07d-12f6f5f0849f">

All traits are looped in the order they are written then calling boot method.

<img width="1440" alt="Screenshot 2024-11-20 at 19 34 25" src="https://github.com/user-attachments/assets/29f993e8-4a60-444e-bcf5-c7e3433112ef">

The boot method calls the observe method directly at this point. But the custom trait has not yet been booted and initialised.

<img width="1440" alt="Screenshot 2024-11-20 at 19 31 11" src="https://github.com/user-attachments/assets/3b99bb5c-1101-41dc-a728-e3f5a92f8314">

---

[https://laravel.com/docs/10.x/eloquent#observers](https://laravel.com/docs/10.x/eloquent#observers)

> To register an observer, you may place the ObservedBy attribute on the corresponding model:

....

> Or, you may manually register an observer by calling the observe method on the model you wish to observe. You may register observers in the boot method of your application's App\Providers\EventServiceProvider service provider:

Laravel documentation says you can use both. But they give different results for this situation.

Thank you in advance

**Comment by crynobone** [comment]
Can you recheck your Laravel version?

**Comment by sdkakcy** [comment]
> Can you recheck your Laravel version?

Yes of course, it is `10.48.17`

**Comment by illambo** [comment]
Hi, I'll also add this #53616 observation (I don't understand if it can be related, if it isn't I apologize).

**Comment by tomschlick** [comment]
So I just ran into this with the [Spatie PrefixedId package](https://github.com/spatie/laravel-prefixed-ids) which uses a `static::creating` call to generate an ID prior to saving. 

https://github.com/spatie/laravel-prefixed-ids/blob/main/src/Models/Concerns/HasPrefixedId.php#L14-L18

If you use that trait, along with an observer defined by the annotation `#[ObservedBy([MyModelObserver::class])]` that also has a `creating` method defined then the trait observer static call gets wiped out.

However if you use the prior method of `Model::observe(MyModelObserver::class)` in the service provider, both that and the trait defined observer work together just fine without conflicts.

I believe its happening because ObservedBy calls `HasEvents::observe()` which generates a new static instance under the hood.
https://github.com/laravel/framework/blob/12.x/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php#L41

I haven't had time to validate this or find a fix, but I just figured I would post this in case someone else can find it before I can dig in.

**Reply by willrowe** [reply]
I've been working on a fix for the issue with it creating a new instance. Creating a new instance in a boot method should not be allowed since it is technically recursive. The only reason it doesn't infinitely loop is because it incorrectly thinks that the class is already booted, even though it is actually still in the process of booting.

**Reply by willrowe** [reply]
@tomschlick I just tried to create a test case for the issue you described, but was not able to reproduce any problems. Looking through the code I also do not see where there would be any problems with this specific case either. Are you able to provide a minimal example that fails to work properly?

**Reply by willrowe** [reply]
Built-in events like `creating` seem to always work fine. The only issue I'm seeing is with custom events, since they would not be registered before the observer is added.

**Reply by tomschlick** [reply]
The issue I encountered seems to be related to events setup during a trait boot with 'bootTraitName' methods. I tried last night to add a fix but haven't had any luck yet.

**Reply by willrowe** [reply]
@tomschlick was that with a custom event or one of the [built-in model events](https://laravel.com/docs/12.x/eloquent#events)?

**Reply by tomschlick** [reply]
It's with the built in creating event. If you look at the trait the spatie package adds (linked above) you'll see it in action. 

So that trait, plus an observer that also triggers the creating event causes the failure for me on L12 with the ObservedBy trait.

**Reply by willrowe** [reply]
@tomschlick as I said, I created a test case for that and built-in model events work properly already. However, If you can provide a simple test case that I can include in a PR, I'm glad to work on a fix for it.

**Comment by willrowe** [comment]
FYI, I submitted #55286 with a very simple fix for the original issue with custom events not working with observers added via attribute. I'm glad to take a look at any other related issues if a simple test case is provided.

**Reply by willrowe** [reply]
@sdkakcy the fix was released in v12.8.0.

**Reply by sdkakcy** [reply]
@willrowe Great news! Well done. Thanks for the information.
